Loading models directly into VRAM, norm calculation on GPUs, broadcasting for ggml_mul

JohannesGaessler opened the pull request:
Currently my implementation of GPU acceleration has inefficient memory management: weights are loaded into RAM and then copied to VRAM. My current goal is to fix this. This PR is a step towards that goal.

Ideally you would directly load parameters from disk into VRAM. However, currently not all parameters can be loaded into VRAM. Between the weight matrices there are two norms per layer. Managing this would make the implementation more complicated. But if all weights in a layer are in VRAM the implementation will be simpler. So this PR implements GPU acceleration for norms (or rather ggml_mul) even though this in and of itself is of rather low priority; for full GPU acceleration we would have needed this eventually though.

On master the norms are first repeated and then multiplied with another tensor. To keep the CUDA code simpler I have extended ggml_mul (for both CPU and CUDA) to allow the broadcasting of values: if `src1` is smaller than `src0` in some dimensions its values are being repeated during the multiplication. That way a repeat CUDA kernel is not needed. The ggml graph is also a little smaller which in theory reduces overhead.

There don't seem to be significant performance differences for generating tokens:

| CPU      | RAM      | GPU      | Model   | -ngl | ms/t master | ms/t ggml_mul broadcast | ms/t ggml_mul broadcast + GPU norms |
|----------|----------|----------|---------|------|-------------|-------------------------|-------------------------------------|
| 3700X    | 3200 MHz | RTX 3090 | 7b q4_0 |    0 |      115.14 |                  115.08 |                              115.00 |
| 3700X    | 3200 MHz | RTX 3090 | 7b q4_0 |   33 |       25.24 |                   24.51 |                               25.30 |
| i5-4570S | 1600 MHz | GTX 1070 | 7b q4_0 |    0 |      205.39 |                  201.77 |                              201.66 |
| i5-4570S | 1600 MHz | GTX 1070 | 7b q4_0 |   33 |       71.33 |                   71.22 |                               72.36 |

JohannesGaessler committed:
Broadcasting for ggml_mul

JohannesGaessler committed:
CUDA kernel for ggml_mul, norms in VRAM

ott2 committed:
define default model path once, sync path with readme (#1366)

ilyakurdyukov committed:
~7% faster Q5_1 AVX2 code (#1477)

slaren commented:
The CUDA code looks fine, but I would suggest some changes to avoid an explosion of `#ifdef GGML_USE_CUBLAS` everywhere:
- Instead of changing the behavior of `ggml_mul` for CUDA only, add the same auto-broadcast behavior to the CPU implementation and remove the ifdefs from llama.cpp. If that's not possible, I would prefer implementing `ggml_repeat` in CUDA. 
- Instead of hooking the CUDA stuff in `ggml_compute_forward_mul_f32`, do it in a generic way in `ggml_compute_forward`. I am thinking of something like this at the start of the function:
  ```C++
  #ifdef GGML_USE_CUBLAS
  if (ggml_cuda_compute_forward(params, tensor)) {
      return;
  }
  #endif
  ```
  Then in the future, when you implement more operations in CUDA, you only need to change `ggml_cuda_compute_forward` without adding more changes to ggml.c.

@ggerganov may disagree about this, wait for his opinion.

JohannesGaessler commented:
There is a CPU implementation for broadcasting of tensors in ggml_mul. It's just that I did not implement a version that utilizes hardware acceleration on CPUs so I thought I would disable it by default.

TheBloke committed:
convert.py: Support models which are stored in a single pytorch_model.bin (#1469)

* Support models in a single pytorch_model.bin

* Remove spurious line with typo

ggerganov commented:
> It's just that I did not implement a version that utilizes hardware acceleration on CPUs so I thought I would disable it by default.

Not sure what you mean by "hardware acceleration on CPUs" - do you mean it is not SIMD-fied? If yes, then it is not a problem.

I think the way you've implemented it is fine and should be enabled by default on the CPU as @slaren recommended

Make sure this does not break the backward computation - run `test-grad0` or `baby-llama` and see if they succeed.

> Instead of hooking the CUDA stuff in ggml_compute_forward_mul_f32, do it in a generic way in ggml_compute_forward

Yes, this is better

JohannesGaessler commented:
I've quickly tried moving the calls for CUDA functions to a new function `ggml_cuda_compute_forward` that is equivalent to `ggml_compute_forward` and then calling that function from `ggml_compute_forward` but this is causing issues with multithreading. The CUDA functions only replace the `GGML_TASK_COMPUTE` parts but are still reliant on the CPU implementation for `GGML_TASK_INIT` and `GGML_TASK_FINALIZE`. Overall it seems like it would be kind of tricky to implement. Can we postpone moving the calls to CUDA functions to `ggml_compute_forward` until we have more than two of them?

Green-Sky commented:
> With the new broadcasting for ggml_mul building on macOS seems to fail. I don't have any Apple devices on which I could debug this; should I just disable it by default again?

you can still check the CI build logs.

JohannesGaessler committed:
GPU weights not in RAM, direct loading with cuFile

JohannesGaessler committed:
GPU weights not in RAM, direct loading with cuFile

Commit:
benchmark-matmul: Print the average of the test results (#1490)

JohannesGaessler commented:
I implemented loading models directly into VRAM and pushed it to this PR; sorry for the feature creep. I'm getting 1.7 t/s for 65b with a 3090 and 32 GB RAM @ 3200 MHz. **This implementation works with EXT4 and NTFS but not BTRFS.**

Models are loaded via cuFile. This allows the GPU to directly access the disk when loading weights. Originally I wanted to have the implementation entirely in `ggml-cuda.cu` but the problem with that is that it would have required a lot of messing with the internals of `llama_model_loader`. So I instead opted to extend `llama_model_loader` and `llama_file`. It's intended to work like this: when building the tensors they get assigned either CPU or CUDA as backend. Then when the actual model is being loaded this information is used to put the data either into RAM or VRAM.

Question: when are there multiple input files? For simplicity I assumed there to be only one.

JohannesGaessler commented:
33b models with 16 GB RAM and 8 GB VRAM still do not seem to be viable. On my headless Linux server I'm ~1 GB short.

ggerganov commented:
> 33b models with 16 GB RAM and 8 GB VRAM still do not seem to be viable. On my headless Linux server I'm ~1 GB short.

When https://github.com/ggerganov/llama.cpp/pull/1508 is ready, it might be just enough to fit it

JohannesGaessler commented:
>I'll fix this over the weekend

Thanks, I'll leave it to you. What I myself still need to do is implement a workaround for Btrfs and fix the info prints for RAM/VRAM usage.

JohannesGaessler commented:
On my Linux server with an i5-4570s, 1600 MHz RAM, and a GTX 1070 when loading 7b q4_0 from a SATA SSD with an empty file system cache:
With cuFile: 15.752 s initialization
With the workaround implementation: 16.949 s initialization.

slaren commented:
Maintaining these changes in llama.cpp or supporting a different backend like opencl is going to be a nightmare. IMO this should be abstracted to the absolute minimum number of dependencies with CUDA and ggml-cuda, and if a dependency cannot be avoided reasonably, it should be clearly isolated from the rest of the code.

Green-Sky commented:
afaik, cufile is also linux only?

Green-Sky commented:
cmake support for cuFile requires at least 3.25
https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#cufile
that's a bit too new

sw committed:
Remove unused n_parts parameter (#1509)

JohannesGaessler commented:
I collected some data with my main machine (RTX 3090, Ryzen 3700X, 32 GB RAM @ 3200 MHz). The test command I used is `time ./main --model /path/to/model --ignore-eos --n_predict 0 --ctx_size 2048 --batch_size 512 --seed 1337 --threads 8 --gpu_layers 99` and I simply measured the runtime. Results:

| Model    | SSD type | 1st initialization? | Runtime cuFile [s] | Runtime workaround [s] |
|----------|----------|---------------------|--------------------|------------------------|
| 7b q4_0  | SATA     | Yes                 |             11.251 |                 12.699 |
| 7b q4_0  | SATA     | No                  |              2.692 |                  4.113 |
| 7b q4_0  | NVME     | Yes                 |              4.471 |                  5.518 |
| 7b q4_0  | NVME     | No                  |              2.740 |                  4.117 |
| 13b q4_0 | SATA     | Yes                 |             20.280 |                 23.149 |
| 13b q4_0 | SATA     | No                  |              3.730 |                  6.492 |
| 13b q4_0 | NVME     | Yes                 |              6.118 |                  9.210 |
| 13b q4_0 | NVME     | No                  |              3.715 |                  6.602 |
| 33b q4_0 | SATA     | Yes                 |             48.123 |                 54.403 |
| 33b q4_0 | SATA     | No                  |              6.394 |                 13.632 |
| 33b q4_0 | NVME     | Yes                 |             12.878 |                 20.442 |
| 33b q4_0 | NVME     | No                  |              6.528 |                 13.607 |

The difference seems to be ~1.5 s for 7b, 3 s for 13b, and 6 s for 33b.

JohannesGaessler commented:
I force-pushed a version based on the feedback. This version keeps any cuFile-related code entirely in `ggml-cuda.cu`. The code for loading weights into VRAM without it is still in `llama.cpp`. This is because it relies on `llama_file` which handles a lot of the OS-specific problems. If the code for loading weights was in `ggml_cuda.cu` instead then either `ggml_cuda.cu` would need to include `llama-util.h` or the relevant code would need to be copy-pasted - I think both of these options are undesirable.

I designed the new version to make it easy to either remove cuFile-related code entirely or to turn it into a compilation option. Opinions regarding this decisions would be appreciated.

The original version can be found [here](https://github.com/JohannesGaessler/llama.cpp/tree/cufile-original).

Green-Sky commented:
> to turn it into a compilation option.

I vote for this.

Has anyone a direct storage system to test this? Either with "Magnum IO GPUDirect Storage" or maybe windows direct storage or something. would be interesting to see.

DannyDaemonic committed:
Fixes #1511 lambda issue for w64devkit (mingw) (#1513)

* Fix for w64devkit and mingw

Green-Sky committed:
make kv_f16 the default for api users (#1517)

rankaiyx commented:
nice!  Is the system with 32G RAM and 11G VRAM expected to run the 65b model?   Or 24G VRAM?

JohannesGaessler commented:
I fixed the compilations issues for MacOS (please someone confirm that the PR produces correct results). Now the issue seems to be that building for Windows fails so if we want to keep cuFile it seems we definitely need to make it a compilation option.

slaren commented on llama.cpp:
```diff
@@ -683,7 +685,7 @@ struct llama_model_loader {
 
         if (use_mmap) {
-            mapping.reset(new llama_mmap(&file_loaders.at(0)->file));
+            mapping.reset(new llama_mmap(&file_loaders.at(0)->file, false));
```
Note that the second parameter to `llama_mmap` here disables madvise/prefetching unconditionally.

JohannesGaessler replied:
Yes, that is something that I want to fix for the final version but I'm not sure about the best approach.

slaren commented on llama.cpp:
```diff
@@ -696,6 +698,9 @@ struct llama_model_loader {
 
         size_t done_size = 0;
         for (llama_load_tensor & lt : tensors_map.tensors) {
+            if (lt.ggml_tensor->backend != GGML_BACKEND_CPU) {
+                continue;
+            }
```
This looks like it may break the `progress_callback`.

JohannesGaessler replied:
The CPU layers are at the front and the GPU layers are at the back. So the current behavior is that progress will jump from some percentage to 100% and then the GPU layers will load without calling the progress callback. Again, the problem is that this is llama functionality that presumably should not be in `ggml-cuda.cu`.

slaren replied:
If I am not missing something, `done_size` is not increased and `data_size` doesn't seem to be updated to remove the GPU layers either, so the progress will never reach 100%.

JohannesGaessler replied:
After the loop in line 716 it's always set to 100%.

slaren commented on llama.cpp:
```diff
@@ -985,77 +970,131 @@ static void llama_model_load_internal(
+                    continue;
+                }
+                size_t actual_size;
+                void * buf = ggml_cuda_pool_malloc(lt.size, &actual_size);
+                void * buf_host = ggml_cuda_host_malloc(lt.size);
```
The `ggml_cuda_pool_malloc` could probably be more efficiently replaced by a `cudaMalloc` directly, no point in keeping these buffers in the pool.

I would also consider keeping this entire branch in `ggml_cuda_load_data_cufile` (maybe renamed to a more generic `ggml_cuda_load_data`).

Also, consider that allocating host memory is sometimes very slow, it may be faster to just use normal memory here.

JohannesGaessler replied:
I can't just put this part of the code into `ggml-cuda.cu` if that is what you mean. Reading from disk is done via `llama_file` (which covers a lot of OS-specific problems) and as such cannot be used in ggml code unless we decide that ggml should include both C++ and llama.

JohannesGaessler replied:
Actually what I just said may be wrong. In `llama_file` C++ std libraries are used but C seems to have the same functionality for working with files. So I think it would be possible without too much work.

JohannesGaessler committed:
fixup! GPU weights not in RAM, direct loading with cuFile

JohannesGaessler commented:
With the tips from @slaren loading without cuFile has become faster. Right now I'm too tired to do extensive testing but the difference compared to cuFile seems insignificant: ~2 s for the first time when loading 33b q4_0, ~0.5 s when loading for the second time. Honestly with a difference like that I think we should just drop cuFile alltogether.

JohannesGaessler commented:
I benchmarked the optimized non-cuFile implementation:

| Model    | SSD type | 1st initialization? | Runtime cuFile [s] | Runtime workaround [s] |
|----------|----------|---------------------|--------------------|------------------------|
| 7b q4_0  | SATA     | Yes                 |             11.251 |                 11.421 |
| 7b q4_0  | SATA     | No                  |              2.692 |                  2.846 |
| 7b q4_0  | NVME     | Yes                 |              4.471 |                  4.132 |
| 7b q4_0  | NVME     | No                  |              2.740 |                  2.838 |
| 13b q4_0 | SATA     | Yes                 |             20.280 |                 20.705 |
| 13b q4_0 | SATA     | No                  |              3.730 |                  4.122 |
| 13b q4_0 | NVME     | Yes                 |              6.118 |                  6.661 |
| 13b q4_0 | NVME     | No                  |              3.715 |                  4.125 |
| 33b q4_0 | SATA     | Yes                 |             48.123 |                 48.126 |
| 33b q4_0 | SATA     | No                  |              6.394 |                  7.726 |
| 33b q4_0 | NVME     | Yes                 |             12.878 |                 14.509 |
| 33b q4_0 | NVME     | No                  |              6.528 |                  7.760 |

I don't think this is enough of a performance uplift to justify the inclusion of cuFile as a compilation option.

rankaiyx commented:
Can two methods be retained? CuFile is limited to certain specifications of NVIDIA cards to maximize performance, and non-cuFile is used to maximize compatibility. Non-cuFile can be adopted by default.

JohannesGaessler committed:
fixup! GPU weights not in RAM, direct loading with cuFile

ggerganov committed:
minor : fix compile warnings

dakennedyd committed:
readme : adds WizardLM to the list of supported models (#1485)

data-angel committed:
main : make reverse prompt option act as a stop token in non-interactive mode (#1032)

* Make reverse prompt option act as a stop token in non-interactive scenarios

* Making requested review changes

* Update gpt_params_parse and fix a merge error

* Revert "Update gpt_params_parse and fix a merge error"

This reverts commit 2bb2ff1748513591ad45b175a75ed1d8089d84c8.

* Update gpt_params_parse and fix a merge error take 2

ejones committed:
examples : add persistent chat (#1495)

* examples : add persistent chat

* examples : fix whitespace

---------

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

ggerganov committed:
tests : add missing header

ggerganov committed:
ggml : use F16 instead of F32 in Q4_0, Q4_1, Q8_0 (#1508)

* ggml : use F16 instead of F32 in Q4_0, Q4_1 and Q8_0

* llama : bump LLAMA_FILE_VERSION to 3

* cuda : update Q4 and Q8 dequantize kernels

* ggml : fix AVX dot products

* readme : update performance table + hot topics

ggerganov committed:
ggml : fix scalar implementation of Q4_1 dot

ggerganov committed:
llama : fix compile warnings in llama_set_state_data()

maximegmd committed:
llama : fix name shadowing and C4146 (#1526)

* Fix name shadowing and C4146

* Fix if macros not using defined when required

* Update llama-util.h

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

* Update llama-util.h

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

* Code style

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

DannyDaemonic committed:
Fix for mingw (#1462)

ggerganov committed:
llama : add llama_init_backend() API (close #1527)

zenixls2 committed:
feature : add blis and other BLAS implementation support (#1502)

* feature: add blis support

* feature: allow all BLA_VENDOR to be assigned in cmake arguments. align with whisper.cpp pr 927

* fix: version detection for BLA_SIZEOF_INTEGER, recover min version of cmake

* Fix typo in INTEGER

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

---------

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

ggerganov committed:
Revert "feature : add blis and other BLAS implementation support (#1502)"

This reverts commit 07e9ace0f9da424d82e75df969642522880feb92.

ggerganov committed:
llama : code style fixes + progress print fix

ggerganov committed:
ggml : ggml_mul better broadcast support

ggerganov committed:
cmake : workarounds for cufile when CMake version < 3.25

ggerganov committed:
Merge branch 'master' into gpu-norms

ggerganov reviewed (changes requested):
I've update the `llama` inference to use broadcasted `ggml_mul()` always.
Cannot do it in `baby-llama` yet because the backward pass is not implemented (added asserts to prevent from anyone using it for now).

I think we have to drop `cufile` - it requires CMake 3.25 and I think currently breaks Windows build for some reason.
Does not seem worth it given the latest benchmarks.

P.S. Hope I didn't mess up something during the rebase that I just did

JohannesGaessler committed:
gg rebase fixup

JohannesGaessler commented:
I removed cuFile and cleaned up the rebase as best as I could. Give me a little more time and I'll clean up the code a little now that I no longer need to consider cuFile.

JohannesGaessler committed:
Loop in llama.cpp, fixed progress callback

JohannesGaessler commented:
I moved the loop for loading weights from `ggml-cuda.cu` to `llama.cpp`. I think this makes the code easier to understand and it also allowed me to fix the progress callback (but now the calculations for that maybe should be deduplicated). As far as I'm concerned I think this can be merged now but I can quickly revise the progress callback code. One thing that I've noticed is that the regular progress indicator is only enabled for mlock because with mmap it's not measurable. With GPU offloading it's always measurable though so I set it to always be enabled.

JohannesGaessler committed:
Attempt clang-tidy fix

ggerganov committed:
llama : fix vram size computation

JohannesGaessler committed:
Add forgotten fclose()

JohannesGaessler commented:
Alright, thanks for the help everyone.
